R/calculate_tdm.R

```diff
@@ -0,0 +1,305 @@
+        0,
+        max(0, (.data$.numerator / .data$.denominator) * .data$monotonic_factor) * 2
+      ),
+      .numerator = NULL,
+      .denominator = NULL
+    ) %>%
+    select(.data$tdm_tech, all_of(groups)) %>%
+    distinct() %>%
+    ungroup()
+}
+
+add_aggregate_tdm <- function(data, t0, groups) {
+  data %>%
+    group_by(!!!rlang::syms(groups)) %>%
+    ungroup(.data$technology) %>%
+    mutate(
+      tdm_sec = .data$plan_carsten * sum(.data$tdm_tech) / sum(.data$plan_carsten),
+      reference_year = t0
+    ) %>%
+    ungroup()
+}
+
+pre_format_data <- function(data, t0, t1, t2, groups) {
+  data %>%
+    filter(.data$year %in% c(t0, t0 + t1, t0 + t2)) %>%
+    group_by(!!!rlang::syms(groups)) %>%
+    add_time_step(t0, t1, t2) %>%
+    select(
+      !!!rlang::syms(groups),
+      .data$time_step,
+      .data$monotonic_factor,
+      scenario_production = .data$scen_alloc_wt_tech_prod,
+      portfolio_production = .data$plan_alloc_wt_tech_prod,
+    ) %>%
+    pivot_wider(
+      names_from = .data$time_step,
+      values_from = all_of(c("scenario_production", "portfolio_production"))
+    ) %>%
+    ungroup()
+}
+
+tdm_prototype <- function() {
+  tibble(
+    technology = character(0),
+    ald_sector = character(0),
+    tdm_tech = numeric(0),
+    tdm_sec = numeric(0),
+    reference_year = integer(0),
+  )
+}
+
+check_calculate_tdm <- function(data, t0, t1, t2, groups) {
+  check_crucial_names(data, crucial_columns())
+
+  check_crucial_years(data, t0, t1, t2, groups)
+
+  check_unique_by_year_and_groups(data, groups)
+
+  invisible(data)
+}
+
+check_unique_by_year_and_groups <- function(data, groups) {
+  req_unique_columns <- c(
+    "scen_alloc_wt_tech_prod",
+    "plan_alloc_wt_tech_prod",
+    "plan_carsten"
+  )
+
+  data <- data %>%
+    group_by(!!!rlang::syms(c("year", groups))) %>%
+    dplyr::summarize(rows_by_year_and_group = dplyr::n()) %>%
+    mutate(rows_are_unique = .data$rows_by_year_and_group == 1)
+
+  ok <- all(data$rows_are_unique)
+  if (!ok) {
+    rlang::abort(
+      "multiple_values_per_year",
+      message = glue(
+        "Data must be unique by year and groups for the following columns:
+      {paste0('`', req_unique_columns, '`', collapse = ', ')} \n
+      Are you sure you have included the correct groupings?"
+      )
+    )
+  }
+
+  invisible(data)
+}
+
+check_crucial_years <- function(data, t0, t1, t2, groups) {
+  crucial_years <- tibble::tibble(crucial_years = c(t0, t0 + t1, t0 + t2))
+
+  missing_crucial_years <- crucial_years %>%
+    left_join(data, by = c(crucial_years = "year")) %>%
+    group_by(!!!rlang::syms(groups)) %>%
+    dplyr::summarize(missing_data = sum(is.na(.data$plan_alloc_wt_tech_prod)))
+
+  ok <- all(missing_crucial_years$missing_data == 0)
+
+  if (!ok) {
+    rlang::abort(
+      "missing_crucial_years",
+      message = glue(
+        "Data must contain all crucial years, by group:
+      {paste0('`', crucial_years, '`', collapse = ', ')} \n
```

Comment:
You may like `glue_collapse()`

``` r
glue::glue_collapse(letters[1:10], ", ", last = " and ", width = 50)
#> a, b, c, d, e, f, g, h, i and j
glue::glue_collapse(letters, ", ", last = " and ", width = 50)
#> a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p,...
```

<sup>Created on 2021-11-12 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>